File: 0021. Merge Two Sorted Lists/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        res = ListNode(0)
        r = res

        p1 = list1
        p2 = list2

        while(p1 and p2):
            if p1.val<p2.val:
                r.next = p1
                r = r.next
                p1 = p1.next

            else:
                r.next = p2
                r = r.next
                p2 = p2.next
        while p1:
            r.next = p1
            r = r.next
            p1 = p1.next

        while p2:
            r.next = p2
            r = r.next
            p2 = p2.next

        return res.next
        

File: 0021. Merge Two Sorted Lists/Solution.java
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode res = new ListNode(0);
        ListNode r = res;

        ListNode p1 = list1;
        ListNode p2 = list2;

        while(p1!=null && p2!=null){
            if(p1.val<p2.val){
                r.next = p1;
                r = r.next;
                p1 = p1.next;
            }
            else {
                r.next = p2;
                r = r.next;
                p2 = p2.next;
            }
        }

        while (p1!=null){
            r.next = p1;
            r = r.next;
            p1 = p1.next;
        }

        while(p2!=null){
            r.next = p2;
            r = r.next;
            p2 = p2.next;
        }

        return res.next;
    }
}